File: apps/php-sdk/src/Models/ParseFile.php
<?php

declare(strict_types=1);

namespace Firecrawl\Models;

use Firecrawl\Exceptions\FirecrawlException;

/**
 * Binary upload payload for the `/v2/parse` endpoint.
 *
 * Supported file extensions: .html, .htm, .pdf, .docx, .doc, .odt, .rtf, .xlsx, .xls
 */
final class ParseFile
{
    private const CONTENT_TYPE_BY_EXTENSION = [
        'pdf' => 'application/pdf',
        'html' => 'text/html',
        'htm' => 'text/html',
        'xhtml' => 'application/xhtml+xml',
        'docx' => 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
        'doc' => 'application/msword',
        'odt' => 'application/vnd.oasis.opendocument.text',
        'rtf' => 'application/rtf',
        'xlsx' => 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        'xls' => 'application/vnd.ms-excel',
    ];

    private function __construct(
        private readonly string $filename,
        private readonly string $content,
        private readonly ?string $contentType,
    ) {}

    /**
     * Build a ParseFile from raw bytes.
     */
    public static function fromBytes(
        string $filename,
        string $content,
        ?string $contentType = null,
    ): self {
        $trimmed = trim($filename);
        if ($trimmed === '') {
            throw new FirecrawlException('filename is required');
        }
        if ($content === '') {
            throw new FirecrawlException('content is required');
        }

        return new self($trimmed, $content, $contentType);
    }

    /**
     * Build a ParseFile by reading a file from disk.
     */
    public static function fromPath(
        string $path,
        ?string $filename = null,
        ?string $contentType = null,
    ): self {
        if ($path === '') {
            throw new FirecrawlException('path is required');
        }
        if (!is_file($path) || !is_readable($path)) {
            throw new FirecrawlException('file path does not exist or is not readable: ' . $path);
        }

        $content = @file_get_contents($path);
        if ($content === false) {
            throw new FirecrawlException('failed to read parse file: ' . $path);
        }

        $resolvedFilename = $filename ?: basename($path);
        $resolvedContentType = $contentType ?: self::guessContentType($resolvedFilename);

        return self::fromBytes($resolvedFilename, $content, $resolvedContentType);
    }

    public function getFilename(): string
    {
        return $this->filename;
    }

    public function getContent(): string
    {
        return $this->content;
    }

    public function getContentType(): ?string
    {
        return $this->contentType;
    }

    private static function guessContentType(string $filename): ?string
    {
        $ext = strtolower(pathinfo($filename, PATHINFO_EXTENSION));
        return self::CONTENT_TYPE_BY_EXTENSION[$ext] ?? null;
    }
}

File: apps/ruby-sdk/lib/firecrawl/models/parse_file.rb
# frozen_string_literal: true

module Firecrawl
  module Models
    # Binary upload payload for the `/v2/parse` endpoint.
    #
    # Supported file extensions: .html, .htm, .pdf, .docx, .doc, .odt, .rtf, .xlsx, .xls
    class ParseFile
      attr_reader :filename, :content, :content_type

      # Build a ParseFile directly.
      #
      # @param filename [String] filename for the upload (e.g., "document.pdf")
      # @param content [String] raw bytes for the file
      # @param content_type [String, nil] optional MIME type hint
      def initialize(filename:, content:, content_type: nil)
        raise ArgumentError, "filename is required" if filename.nil? || filename.to_s.strip.empty?
        raise ArgumentError, "content is required" if content.nil? || content.bytesize.zero?

        @filename = filename.to_s.strip
        @content = content.to_s
        @content_type = content_type
      end

      # Build a ParseFile by reading a file from disk.
      #
      # @param path [String] absolute or relative path to the file
      # @param filename [String, nil] optional override for the upload filename
      # @param content_type [String, nil] optional MIME type hint
      # @return [ParseFile]
      def self.from_path(path, filename: nil, content_type: nil)
        raise ArgumentError, "path is required" if path.nil? || path.to_s.strip.empty?
        unless File.file?(path)
          raise ArgumentError, "file path does not exist: #{path}"
        end

        content = File.binread(path)
        resolved_filename = filename || File.basename(path)
        resolved_content_type = content_type || guess_content_type(resolved_filename)
        new(filename: resolved_filename, content: content, content_type: resolved_content_type)
      end

      # @api private
      def self.guess_content_type(filename)
        ext = File.extname(filename).downcase
        {
          ".pdf" => "application/pdf",
          ".html" => "text/html",
          ".htm" => "text/html",
          ".xhtml" => "application/xhtml+xml",
          ".docx" => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
          ".doc" => "application/msword",
          ".odt" => "application/vnd.oasis.opendocument.text",
          ".rtf" => "application/rtf",
          ".xlsx" => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
          ".xls" => "application/vnd.ms-excel",
        }[ext]
      end
    end
  end
end
